c++/src/kj/compat/http.c++

```diff
@@ -4093,6 +4093,103 @@ private:
   kj::ArrayPtr<byte> leftover;
 };
 
+class AsyncIoStreamWithDeferredInput final: public kj::AsyncIoStream {
+  // This AsyncIoStream immediately allows write to the given wrapped AsyncIoStream
+  // but will delay reads until after a guard promise is resolved. When the guard
+  // promise resolves, it may provide a released buffer that will be read out first.
+  // The primary use case for this impl is to support pipelined CONNECT calls which
+  // optimistically allow outbound writes to happen while establishing the CONNECT
+  // tunnel has not yet been completed. If the guard promise rejects, the stream
+  // is permanently errored and existing pending calls (reads and writes) are canceled.
+public:
+  AsyncIoStreamWithDeferredInput(
+      kj::Own<kj::AsyncIoStream> inner,
+      kj::Promise<HttpInputStreamImpl::ReleasedBuffer> guard)
+      : inner(kj::mv(inner)),
+        guard(handleGuard(kj::mv(guard))) {}
+
+  // AsyncInputStream
+  Promise<size_t> tryRead(void* buffer, size_t minBytes, size_t maxBytes) override {
+    KJ_IF_MAYBE(exception, permanentException) {
+      return kj::cp(*exception);
+    }
+    return guard.addBranch().then([this, buffer, minBytes, maxBytes] {
+      return inner->tryRead(buffer, minBytes, maxBytes);
+    });
+  }
+
+  Maybe<uint64_t> tryGetLength() override {
+    return nullptr;
+  }
+
+  kj::Promise<uint64_t> pumpTo(AsyncOutputStream& output, uint64_t amount) override {
+    KJ_IF_MAYBE(exception, permanentException) {
+      return kj::cp(*exception);
+    }
+    return guard.addBranch().then([this, &output, amount] {
+      return inner->pumpTo(output, amount);
+    });
+  }
+
+  // AsyncOutputStream
+
+  void shutdownWrite() override {
+    KJ_IF_MAYBE(exception, permanentException) {
+      throw *exception;
```

Comment:
We don't use bare `throw` in KJ, use `kj::throwRecoverableException` or `kj::throwFatalException` instead (these two functions do the same thing when compiling with exceptions enabled, but different things when exceptions are disabled).